### PR TITLE
Optimize JTF Collection size in ReentrantSemaphore

### PR DESCRIPTION
**Pull request by jviau**
The pattern of:
``` CSharp
this.joinableTaskFactory.RunAsync(async () =>
{
    using (this.joinableTaskCollection.Join())
    using (await this.semaphore.EnterAsync())
    {
        await externalOperation();
    }
});
```

is problematic in that it explodes the size of the joinable task collection, leading to a large dependency graph and serious performance issues. This PR proposes a fix by:

1. Capture synchronization context
2. Await semaphore _without_ using `JTF.RunAsync`. This requires `ConfigureAwaitRunInline()` (hence step 1).
3. `RunAsync` and restore synchronization context.

Let me know what you think of this change, and if the general concept looks good I will apply it to all of `ReentrantSemaphore` and add unit tests as appropriate.

Another thing to consider, is to change `AsyncSemaphore` to use the `SynchronizationContextAwaiter` directly, but I think that would be a breaking change.

/cc: @lifengl

### CODE REVIEW

File: src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs [hunk 1]
```diff
@@ -408,35 +408,53 @@ internal StackSemaphore(int initialCount, JoinableTaskContext joinableTaskContex
                     this.reentrantCount.Value = reentrantStack = new Stack<StrongBox<AsyncSemaphore.Releaser>>(capacity: 2);
                 }
 
-                await this.ExecuteCoreAsync(async delegate
+                // For performance reasons, we want to minimize the number of Joins performed, and also keep the size of the
+                // JoinableCollection to a minimum. This also means awaiting on the semaphore outside of a JTF.RunAsync. This requires
+                // us to not ConfigureAwait(true) on the semaphore. However, that breaks out contract. To fix this, we capture the
+                // synchronization context and then restore it after acquiring the semaphore.
```
[the remark this answers is not in the document]
Comment (jviau) [reply src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: Can you clarify what you mean here? Is the issue with the `SynchronizationContextAwaiter`, or with my usage of it here? Are you wanting `SynchronizationContext.GetAwaiter` to work on a null `SynchronizationContext`?

File: src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs [hunk 2]
```diff
@@ -408,35 +408,53 @@ internal StackSemaphore(int initialCount, JoinableTaskContext joinableTaskContex
...
                     using (this.joinableTaskCollection?.Join())
                     {
-                        AsyncSemaphore.Releaser releaser = reentrantStack.Count == 0 ? await this.semaphore.EnterAsync(cancellationToken).ConfigureAwait(true) : default;
-                        var pushedReleaser = new StrongBox<AsyncSemaphore.Releaser>(releaser);
-                        lock (reentrantStack)
-                        {
-                            reentrantStack.Push(pushedReleaser);
-                        }
+                        // Use ConfiguredAwaitRunInline() as ConfigureAwait(true) will deadlock due to not being inside a JTF.RunAsync().
+                        releaser = await this.semaphore.EnterAsync(cancellationToken).ConfigureAwaitRunInline();
```
[the remark this answers is not in the document]
Comment (jviau) [reply src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: What is the fix here then? Is it related to your other comment - if the captured sync context is null, switch to `TaskScheduler.Default`?

File: src/Microsoft.VisualStudio.Threading/AwaitExtensions.cs [hunk 3]
```diff
@@ -852,5 +863,45 @@ public void OnCompleted(Action continuation)
...
+            /// </summary>
+            public bool IsCompleted => this.context == SynchronizationContext.Current;
+
+            /// <summary>
+            /// Posts the callback to the supplied synchronization context.
+            /// </summary>
+            /// <param name="continuation">The action to continue</param>
+            public void OnCompleted(Action continuation)
+            {
+                this.context.Post(PostCallback, continuation);
```
[the remark this answers is not in the document]
Comment (jviau) [reply src/Microsoft.VisualStudio.Threading/AwaitExtensions.cs]: This behavior really is up to the individual `SynchronizationContext`. For example, the [XUnit AsyncTestSyncContext](https://github.com/xunit/xunit/blob/bccfcccf26b2c63c90573fe1a17e6572882ef39c/src/xunit.execution/Sdk/AsyncTestSyncContext.cs) becomes null after any post to it. So be it this extension method, or even an `await Task.Yield()`, `SynchronizationContext.Current` becomes null. As for `DispatcherSynchronizationContext`, the await extension works. But a `Task.Yield()`, even without using the await extension, returns on a different `DispatcherSynchronizationContext`.

File: src/Microsoft.VisualStudio.Threading/AwaitExtensions.cs [hunk 4]
```diff
@@ -49,6 +49,17 @@ public static TaskSchedulerAwaitable SwitchTo(this TaskScheduler scheduler, bool
             return new TaskSchedulerAwaitable(scheduler, alwaysYield);
         }
 
+        /// <summary>
+        /// Gets an awaitable that schedules continuations on the specified synchronization context.
+        /// </summary>
+        /// <param name="context">The synchronization context used to execute continuations.</param>
+        /// <returns>An awaitable.</returns>
+        public static SynchronizationContextAwaiter GetAwaiter(this SynchronizationContext context)
```
[the remark this answers is not in the document]
Comment (AArnott) [reply src/Microsoft.VisualStudio.Threading/AwaitExtensions.cs]: As discussed yesterday, let's keep this internal. Folks who await a SynchronizationContext (or a TaskScheduler) may deadlock on shutdown because they're not leveraging JTF. So it's not a pattern we want to encourage.

---
In reply to: [200798277](https://github.com/Microsoft/vs-threading/pull/313#discussion_r200798277) [](ancestors = 200798277)

File: src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (AArnott) [reply src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: My concern here is that you're trying to execute the user's delegate on their SynchronizationContext only. If that is null, the behavior they would expect is that you then fallback to using their TaskScheduler.Current.

---
In reply to: [201100340](https://github.com/Microsoft/vs-threading/pull/313#discussion_r201100340) [](ancestors = 201100340)

[the remark this answers is not in the document]
Comment (jviau) [reply src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: Ah okay, that is a simple fix - if it is null I will `await TaskScheduler.Default;`

File: src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs [hunk 5]
```diff
@@ -274,19 +274,41 @@ internal NotRecognizedSemaphore(int initialCount, JoinableTaskContext joinableTa
             {
                 Requires.NotNull(operation, nameof(operation));
 
+                // Note: this code is duplicated and not extracted to minimize allocating extra async state machines.
+                // For performance reasons, we want to minimize the number of Joins performed, and also keep the size of the
+                // JoinableCollection to a minimum. This also means awaiting on the semaphore outside of a JTF.RunAsync. This requires
+                // us to not ConfigureAwait(true) on the semaphore. However, that breaks our contract of resuming on the sync context
+                // you started on. To partially fix this, we will at least resume you on the main thread or thread pool.
+                AsyncSemaphore.Releaser releaser = default;
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: > = default [](start = 48, length = 10)

Can you remove this initializer? We unconditionally assign a value to the field further down. If we stopped doing that or added a condition to it, it might be by accident and we'd want the compiler to call it out. This initial value doesn't offer us anything, and it would suppress that compiler warning.
Same everywhere else it appears.

File: src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs [hunk 6]
```diff
@@ -274,19 +274,41 @@ internal NotRecognizedSemaphore(int initialCount, JoinableTaskContext joinableTa
...
-                        try
-                        {
-                            await operation().ConfigureAwaitRunInline();
-                        }
-                        finally
-                        {
-                            DisposeReleaserNoThrow(releaser);
-                        }
+                        // Return to the main thread if we started there.
+                        await this.joinableTaskFactory.SwitchToMainThreadAsync();
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: >() [](start = 78, length = 2)

We should pass in `cancellationToken`.
We should also be prepared to release the semaphore if it were to throw. So I believe this whole if/else block should be moved into the try below.

File: src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs [hunk 7]
```diff
@@ -274,19 +274,41 @@ internal NotRecognizedSemaphore(int initialCount, JoinableTaskContext joinableTa
             {
                 Requires.NotNull(operation, nameof(operation));
 
+                // Note: this code is duplicated and not extracted to minimize allocating extra async state machines.
+                // For performance reasons, we want to minimize the number of Joins performed, and also keep the size of the
+                // JoinableCollection to a minimum. This also means awaiting on the semaphore outside of a JTF.RunAsync. This requires
+                // us to not ConfigureAwait(true) on the semaphore. However, that breaks our contract of resuming on the sync context
+                // you started on. To partially fix this, we will at least resume you on the main thread or thread pool.
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: We don't really have a contract yet, as this is a new feature. So rather than say it breaks a contract, let's just state what our contract is. And for that matter, it would be good to express this in the xml doc comments (on the base class) for this method.

File: src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs [hunk 8]
```diff
@@ -142,6 +141,9 @@ public static ReentrantSemaphore Create(int initialCount = 1, JoinableTaskContex
 
         /// <summary>
         /// Executes a given operation within the semaphore.
+        /// <paramref name="operation"/> is not guaranteed to be executed on the context this was originally called on. If a
+        /// <see cref="JoinableTaskContext"/> was supplied to the ctor, <paramref name="operation"/> will at least be called on
+        /// the main thread if it was originaly on the main thread.
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: >originaly [](start = 38, length = 9)

typo: originaly -> originally

File: src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs [hunk 9]
```diff
@@ -274,19 +276,48 @@ internal NotRecognizedSemaphore(int initialCount, JoinableTaskContext joinableTa
...
                 Requires.NotNull(operation, nameof(operation));
 
+                // Note: this code is duplicated and not extracted to minimize allocating extra async state machines.
+                // For performance reasons in the JTF enabled scenario, we want to minimize the number of Joins performed, and also
+                // keep the size of the JoinableCollection to a minimum. This also means awaiting on the semaphore outside of a
+                // JTF.RunAsync. This requires us to not ConfigureAwait(true) on the semaphore. However, that prevents us from
+                // resuming on the correct sync context. To partially fix this, we will at least resume you on the main thread or
+                // thread pool.
+                AsyncSemaphore.Releaser releaser;
+                bool resumeOnMainThread = this.joinableTaskCollection?.Context?.IsOnMainThread == true;
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: > == true [](start = 94, length = 8)

nit: Drop this

File: src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs [hunk 10]
```diff
@@ -274,19 +276,48 @@ internal NotRecognizedSemaphore(int initialCount, JoinableTaskContext joinableTa
...
+                                await this.joinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                            }
+                            else
+                            {
+                                await TaskScheduler.Default;
+                            }
+
+                            // Always yield to prevent running on the stack that released the semaphore.
+                            await Task.Yield();
                         }
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: We're so nearly there. But I just realized that this change introduces a yield even when there was no semaphore contention for the caller (we should have a test for this).
If you revise your `release = usingJtf` assignment. to *not* await directly, but simply assign the unawaited result to a local variable, then we can call `IsCompleted` on that ourselves and save another local indicating whether it was completed (telling us there was no contention and thus no yielding). We then of course will need to await that variable that stores the `EnterAsync` result. We can then condition *this* `await Task.Yield`  based on whether `!IsCompleted` from earlier.

File: src/Microsoft.VisualStudio.Threading.Tests/ReentrantSemaphoreJTFTests.cs [hunk 11]
```diff
@@ -62,6 +62,57 @@ async delegate
...
 
+    [Theory]
+    [MemberData(nameof(AllModes))]
+    public void SemaphoreDoesNotDeadlockReturningToMainThread(ReentrantSemaphore.ReentrancyMode mode)
+    {
+        this.semaphore = this.CreateSemaphore(mode);
+        this.ExecuteOnDispatcher(
+            async () =>
+            {
+                var semaphoreAquired = new AsyncManualResetEvent();
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading.Tests/ReentrantSemaphoreJTFTests.cs]: >semaphoreAquired [](start = 20, length = 16)

sp: semaphoreAquired -> semaphoreAcquired

File: src/Microsoft.VisualStudio.Threading.Tests/ReentrantSemaphoreTestBase.cs [hunk 12]
```diff
@@ -180,7 +180,8 @@ public void Uncontested(int initialCount)
                 // Fill the semaphore to its capacity
                 for (int j = 0; j < initialCount; j++)
                 {
-                    operations[j] = this.semaphore.ExecuteAsync(() => releasers[j].WaitAsync(), this.TimeoutToken);
+                    int k = j; // Capture j, as it will increment
+                    operations[j] = this.semaphore.ExecuteAsync(() => releasers[k].WaitAsync(), this.TimeoutToken);
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading.Tests/ReentrantSemaphoreTestBase.cs]: :smile:

File: src/Microsoft.VisualStudio.Threading.Tests/ReentrantSemaphoreTestBase.cs [hunk 13]
```diff
@@ -227,23 +228,24 @@ public void ExitInAnyOrder_ExitInAcquisitionOrder()
...
+                innerOperation = EnterAndUseSemaphoreAsync(innerReleaser, outerReleaser);
+                await outerReleaser;
             });
             innerReleaser.Set();
             await innerOperation;
             Assert.Equal(1, this.semaphore.CurrentCount);
         });
 
-        async Task EnterAndUseSemaphoreAsync(AsyncManualResetEvent releaseEvent)
+        async Task EnterAndUseSemaphoreAsync(AsyncManualResetEvent releaseEvent, AsyncManualResetEvent signal)
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading.Tests/ReentrantSemaphoreTestBase.cs]: >signal [](start = 103, length = 6)

Given we take two of these, can we name this more distinctively? Perhaps `enteredEvent`?

Comment (AArnott) [reply src/Microsoft.VisualStudio.Threading.Tests/ReentrantSemaphoreTestBase.cs]: Also, I'd rather reorder these parameters so that they are used in-order.

---
In reply to: [201717015](https://github.com/Microsoft/vs-threading/pull/313#discussion_r201717015) [](ancestors = 201717015)

File: src/Microsoft.VisualStudio.Threading.Tests/ReentrantSemaphoreTestBase.cs [hunk 14]
```diff
@@ -227,23 +228,24 @@ public void ExitInAnyOrder_ExitInAcquisitionOrder()
...
             var innerReleaser = new AsyncManualResetEvent();
             Task innerOperation = null;
-            await this.semaphore.ExecuteAsync(delegate
+            await this.semaphore.ExecuteAsync(async delegate
             {
-                innerOperation = EnterAndUseSemaphoreAsync(innerReleaser);
-                return TplExtensions.CompletedTask;
+                innerOperation = EnterAndUseSemaphoreAsync(innerReleaser, outerReleaser);
+                await outerReleaser;
             });
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading.Tests/ReentrantSemaphoreTestBase.cs]: Ah, these changes were probably necessary due to the unconditional `await Task.Yield();` we added. Fixing that as I mention at that place in the code should allow you to revert these changes.

File: src/Microsoft.VisualStudio.Threading.Tests/ReentrantSemaphoreTestBase.cs [hunk 15]
```diff
@@ -447,14 +449,16 @@ public void Stack_ViolationCaughtAtBothSites()
...
                 {
+                    release3.Set();
                     await release2;
                 });
 
                 Assert.Equal(0, this.semaphore.CurrentCount);
 
-                await release1;
+                await Task.WhenAll(release1.WaitAsync(), release3.WaitAsync());
             });
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading.Tests/ReentrantSemaphoreTestBase.cs]: Same here: these test changes can probably be reverted.

File: src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs (hunk 9 above)
Comment (jviau) [reply src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: I changed it up, but the type is a `bool?`, using `== true` coalesces to `bool`.

File: src/Microsoft.VisualStudio.Threading.Tests/ReentrantSemaphoreTestBase.cs (hunk 14 above)
Comment (jviau) [reply src/Microsoft.VisualStudio.Threading.Tests/ReentrantSemaphoreTestBase.cs]: Is that behavior you want the test to enforce though? That it does not yield at all when there is no semaphore contention?

File: src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs [hunk 16]
```diff
@@ -143,7 +142,12 @@ public static ReentrantSemaphore Create(int initialCount = 1, JoinableTaskContex
         /// <summary>
         /// Executes a given operation within the semaphore.
         /// </summary>
-        /// <param name="operation">The delegate to invoke once the semaphore is entered.</param>
+        /// <param name="operation">
+        /// The delegate to invoke once the semaphore is entered. Not guaranteed to be executed on the context this was originally called on.
+        /// If a <see cref="JoinableTaskContext"/> was supplied to the constructor, this delegate will execute on the main thread if  is
+        /// invoked on the main thread, otherwise it will be invoked on the threadpool. When no <see cref="JoinableTaskContext"/> is supplied
+        /// to the constructor, this delegate will execute on the caller's context or the task scheduler.
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: >context or the task scheduler [](start = 75, length = 29)

the caller's task scheduler is part of their context. So I'd leave it at just "the caller's context" or "the caller's SynchronizationContext or TaskScheduler".

File: src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs [hunk 17]
```diff
@@ -143,7 +142,12 @@ public static ReentrantSemaphore Create(int initialCount = 1, JoinableTaskContex
         /// <summary>
         /// Executes a given operation within the semaphore.
         /// </summary>
-        /// <param name="operation">The delegate to invoke once the semaphore is entered.</param>
+        /// <param name="operation">
+        /// The delegate to invoke once the semaphore is entered. Not guaranteed to be executed on the context this was originally called on.
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: > Not guaranteed to be executed on the context this was originally called on. [](start = 65, length = 76)

I would drop this sentence, since it *is* guaranteed in some conditions, which you fully describe after this sentence.

File: src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs (hunk 9 above)
Comment (AArnott) [reply src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: Ah, in that case I guess what you had was fine. Sorry for the noise. You can keep it the new way too. 

---
In reply to: [201787699](https://github.com/Microsoft/vs-threading/pull/313#discussion_r201787699) [](ancestors = 201787699)

File: src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs [hunk 18]
```diff
@@ -274,19 +278,60 @@ internal NotRecognizedSemaphore(int initialCount, JoinableTaskContext joinableTa
...
+                // Note: this code is duplicated and not extracted to minimize allocating extra async state machines.
+                // For performance reasons in the JTF enabled scenario, we want to minimize the number of Joins performed, and also
+                // keep the size of the JoinableCollection to a minimum. This also means awaiting on the semaphore outside of a
+                // JTF.RunAsync. This requires us to not ConfigureAwait(true) on the semaphore. However, that prevents us from
+                // resuming on the correct sync context. To partially fix this, we will at least resume you on the main thread or
+                // thread pool.
+                AsyncSemaphore.Releaser releaser;
+                bool usingJtf = this.joinableTaskFactory != null;
+                bool resumeOnMainThread = usingJtf ? this.joinableTaskCollection.Context.IsOnMainThread : false;
+                bool mustYield = false;
```
Comment (MTBandolier) [inline src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: These are identical in each concrete class and feel like they could be properties on the base class.  Is there any reason you didn't implement them that way?

File: src/Microsoft.VisualStudio.Threading.Tests/ReentrantSemaphoreTestBase.cs [hunk 19]
```diff
@@ -227,7 +228,7 @@ public void ExitInAnyOrder_ExitInAcquisitionOrder()
         this.semaphore = this.CreateSemaphore(ReentrantSemaphore.ReentrancyMode.Freeform);
         this.ExecuteOnDispatcher(async delegate
         {
-            var releaser1 = new AsyncManualResetEvent();
+            var outerReleaser = new AsyncManualResetEvent();
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading.Tests/ReentrantSemaphoreTestBase.cs]: >outerReleaser [](start = 16, length = 13)

So... this change seems to be to a variable that no one uses. Shall we revert the change, or simply delete the line?

File: src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs (hunk 18 above)
Comment (jviau) [reply src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: They are all locals. I could move `usingJtf` to be a readonly bool though.

Comment (MTBandolier) [reply src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: (I was commenting in CodeFlow and didn't realize that the highlighted lines don't seem to show up very well.)

I was referring to ```usingJtf``` and ```resumeOnMainThread```.  Those definitions are identical in all four concrete types.  You could make ```usingJtf``` a ```readonly bool```, but I'm not sure it's worth the extra field.  I was envisioning something like this in ```ReentrantSemaphore```:

```cs
protected bool UsingJtf => this.joinableTaskFactory != null;
protected bool ResumeOnMainThread => this.UsingJtf ? this.joinableTaskCollection.Context.IsOnMainThread : false;
```

Not a huge deal, though.

Comment (jviau) [reply src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs]: I will extract `UsingJtf` to a property like you suggested, that is a good idea. `ResumeOnMainThread` is not much of a benefit as it must captured at that point in time. I think capturing a property at that point will be less descriptive of what I am trying to achieve there.

**Comment by AArnott** [comment]
@jviau @MTBandolier Is it clear who is sending this through escrow?

**Comment by MTBandolier** [comment]
If you can give me the change to Default.config, I can include this change with my escrow bug fix.

**Comment by AArnott** [comment]
@mtbandolier Thanks, but before we have the change for default.config we'll have to merge the change into vs-threading's v15.8 branch, so we'll need escrow approval earlier than yours.
